{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4f1c464d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tskit 0.4.1\n",
      "tsinfer 0.2.3.dev9+gc8568d5\n",
      "msprime 1.1.1\n",
      "stdpopsim 0.1.2\n",
      "cyvcf2 0.30.14\n"
     ]
    }
   ],
   "source": [
    "from copy import deepcopy\n",
    "import math\n",
    "import numpy as np\n",
    "import os\n",
    "import random\n",
    "import sys\n",
    "import time\n",
    "\n",
    "from IPython.display import SVG\n",
    "\n",
    "sys.path.append(\"../modules/\")\n",
    "import mask_genotype\n",
    "import parse_vcf\n",
    "\n",
    "import demes\n",
    "import tsinfer\n",
    "from tsinfer import make_ancestors_ts\n",
    "import tskit\n",
    "import msprime\n",
    "import cyvcf2\n",
    "\n",
    "print(f\"tskit {tskit.__version__}\")\n",
    "print(f\"tsinfer {tsinfer.__version__}\")\n",
    "print(f\"msprime {msprime.__version__}\")\n",
    "print(f\"stdpopsim {stdpopsim.__version__}\")\n",
    "print(f\"cyvcf2 {cyvcf2.__version__}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e64d1b6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_sample_data_to_vcf(sample_data,\n",
    "                             individuals,\n",
    "                             samples,\n",
    "                             ploidy_level,\n",
    "                             mask,\n",
    "                             out_vcf_file,\n",
    "                             contig_id,\n",
    "                             sequence_length_max = 1e12):\n",
    "    \"\"\"\n",
    "    Fields:\n",
    "    CHROM contig_id\n",
    "    POS row index in genotype_matrix\n",
    "    ID .\n",
    "    REF ancestral allele\n",
    "    ALT derived allele(s)\n",
    "    QUAL .\n",
    "    FILTER PASS\n",
    "    INFO\n",
    "    FORMAT GT\n",
    "    individual 0\n",
    "    individual 1\n",
    "    ...\n",
    "    individual n - 1; n = number of individuals\n",
    "    \"\"\"\n",
    "    CHROM = contig_id\n",
    "    ID = '.'\n",
    "    QUAL = '.'\n",
    "    FILTER = 'PASS'\n",
    "    FORMAT = 'GT'\n",
    "    \n",
    "    assert ploidy_level == 1 or ploidy_level == 2,\\\n",
    "        f\"Specified ploidy_level {ploidy_level} is not recognized.\"\n",
    "    \n",
    "    assert ploidy_level * len(individuals) == len(samples),\\\n",
    "        f\"Some individuals may not have the same ploidy level of {ploidy_level}.\"\n",
    "    \n",
    "    # Assume that both sample and individual ids are ordered the same way.\n",
    "    #individual_id_map = np.repeat(individuals, 2)\n",
    "    \n",
    "    header  = \"##fileformat=VCFv4.2\\n\"\\\n",
    "            + \"##source=tskit \" + tskit.__version__ + \"\\n\"\\\n",
    "            + \"##INFO=<ID=AA,Number=1,Type=String,Description=\\\"Ancestral Allele\\\">\\n\"\\\n",
    "            + \"##FORMAT=<ID=GT,Number=1,Type=String,Description=\\\"Genotype\\\">\\n\"\n",
    "    header += \"##contig=<ID=\" + contig_id + \",\" + \"length=\" + str(int(ts.sequence_length)) + \">\\n\"\n",
    "    header += \"\\t\".join(['#CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT']\\\n",
    "                        + [\"s\" + str(x) for x in individuals])\n",
    "    \n",
    "    with open(out_vcf_file, \"w\") as vcf:\n",
    "        vcf.write(header + \"\\n\")\n",
    "        for i, variant in enumerate(ts.variants()):\n",
    "            site_id = variant.site.id\n",
    "            POS = int(np.round(variant.site.position))\n",
    "            if POS > sequence_length_max:\n",
    "                break\n",
    "            # Since the tree sequence was produced using simulation,\n",
    "            #    there's no reference sequence other than the ancestral sequence.\n",
    "            REF = variant.site.ancestral_state\n",
    "            alt_alleles = list(set(variant.alleles) - {REF})\n",
    "            AA = variant.site.ancestral_state\n",
    "            ALT = \",\".join(alt_alleles) if len(alt_alleles) > 0 else \".\"\n",
    "            INFO = \"AA\" + \"=\" + AA\n",
    "            record = [str(x)\n",
    "                      for x\n",
    "                      in [CHROM, POS, ID, REF, ALT, QUAL, FILTER, INFO, FORMAT]]\n",
    "            \n",
    "            for j in individuals:\n",
    "                #sample_ids = [samples[x]\n",
    "                #              for x\n",
    "                #              in np.where(individual_id_map == j)[0].tolist()]\n",
    "                #genotype = \"|\".join([str(variant.genotypes[k])\n",
    "                #                     for k\n",
    "                #                     in sample_ids])\n",
    "                if ploidy_level == 1:\n",
    "                    genotype = str(variant.genotypes[j])\n",
    "                else:\n",
    "                    genotype = str(variant.genotypes[2 * j]) + \"|\" + str(variant.genotypes[2 * j + 1])\n",
    "                    \n",
    "                if mask is not None and mask.query_position(individual = j, position = POS) == True:\n",
    "                    if ploidy_level == 1:\n",
    "                        genotype = '.'\n",
    "                    else:\n",
    "                        genotype = '.|.' # Or \"./.\"\n",
    "                record += [genotype]\n",
    "                \n",
    "            vcf.write(\"\\t\".join(record) + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "aa396407",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sourced and modified from:\n",
    "# https://tsinfer.readthedocs.io/en/latest/tutorial.html#data-example\n",
    "def get_chromosome_length(vcf):\n",
    "    assert len(vcf.seqlens) == 1\n",
    "    return vcf.seqlens[0]\n",
    "\n",
    "\n",
    "def add_populations(vcf,\n",
    "                    samples):\n",
    "    \"\"\"\n",
    "    TODO\n",
    "    \"\"\"\n",
    "    pop_ids = [sample_name[0] for sample_name in vcf.samples]\n",
    "    pop_codes = np.unique(pop_ids)\n",
    "    pop_lookup = {}\n",
    "    for p in pop_codes:\n",
    "        pop_lookup[p] = samples.add_population(metadata = {\"name\" : p})\n",
    "    return [pop_lookup[pop_id] for pop_id in pop_ids]\n",
    "\n",
    "\n",
    "def add_individuals(vcf,\n",
    "                    samples,\n",
    "                    ploidy_level,\n",
    "                    populations):\n",
    "    for name, population in zip(vcf.samples, populations):\n",
    "        samples.add_individual(ploidy = ploidy_level,\n",
    "                               metadata = {\"name\": name},\n",
    "                               population = population)\n",
    "\n",
    "\n",
    "def add_sites(vcf,\n",
    "              samples,\n",
    "              ploidy_level,\n",
    "              warn_monomorphic_sites = False):\n",
    "    \"\"\"\n",
    "    Read the sites in the VCF and add them to the samples object,\n",
    "    reordering the alleles to put the ancestral allele first,\n",
    "    if it is available.\n",
    "    \"\"\"\n",
    "    assert ploidy_level == 1 or ploidy_level == 2,\\\n",
    "        f\"ploidy_level {ploidy_level} is not recognized.\"\n",
    "    \n",
    "    pos = 0\n",
    "    for variant in vcf:\n",
    "        # Check for duplicate site positions.\n",
    "        if pos == variant.POS:\n",
    "            raise ValueError(\"Duplicate positions for variant at position\", pos)\n",
    "        else:\n",
    "            pos = variant.POS\n",
    "        # Check that the genotypes are phased.\n",
    "        #if any([not phased for _, _, phased in variant.genotypes]):\n",
    "        #    raise ValueError(\"Unphased genotypes for variant at position\", pos)\n",
    "        alleles = [variant.REF] + variant.ALT # Exactly as in the input VCF file.\n",
    "        if warn_monomorphic_sites:\n",
    "            if len(set(alleles) - {'.'}) == 1:\n",
    "                print(f\"Monomorphic site at {pos}\")\n",
    "        ancestral = variant.INFO.get(\"AA\", variant.REF) # Dangerous action!!!\n",
    "        # Ancestral state must be first in the allele list.\n",
    "        ordered_alleles = [ancestral] + list(set(alleles) - {ancestral})\n",
    "        # Create an index mapping from the input VCF to tsinfer input.\n",
    "        allele_index = {\n",
    "            old_index: ordered_alleles.index(allele)\n",
    "            for old_index, allele in enumerate(alleles)\n",
    "        }\n",
    "        # When genotype is missing...\n",
    "        if variant.num_unknown > 0:\n",
    "            allele_index[-1] = tskit.MISSING_DATA\n",
    "            ordered_alleles += [None]\n",
    "        # Map original allele indexes to their indexes in the new alleles list.\n",
    "        genotypes = [\n",
    "            allele_index[old_index]\n",
    "            for row in variant.genotypes # cyvcf2 uses -1 to indicate missing data.\n",
    "            for old_index in row[0:ploidy_level] # Each is a 3-tuple (allele 1, allele 2, is phased?).\n",
    "        ]\n",
    "        samples.add_site(pos,\n",
    "                         genotypes = genotypes,\n",
    "                         alleles = ordered_alleles)\n",
    "\n",
    "\n",
    "def create_sample_data_from_vcf_file(vcf_file):\n",
    "    vcf = cyvcf2.VCF(vcf_file,\n",
    "                     gts012 = False, # 0=HOM_REF, 1=HET, 2=UNKNOWN, 3=HOM_ALT\n",
    "                     strict_gt = True)\n",
    "    with tsinfer.SampleData(\n",
    "        sequence_length = get_chromosome_length(vcf)\n",
    "    ) as samples:\n",
    "        populations = add_populations(vcf, samples)\n",
    "        add_individuals(vcf, samples, ploidy_level, populations)\n",
    "        add_sites(vcf, samples, ploidy_level)\n",
    "    return(samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "23154dcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ts_with_discretized_coordinates(ts):\n",
    "    ts_tables = ts.dump_tables()\n",
    "    ts_tables.sites.position = np.round(ts_tables.sites.position)\n",
    "    ts_tables.deduplicate_sites()\n",
    "    ts_tables.sort()\n",
    "    ts_tables.build_index()\n",
    "    ts_tables.compute_mutation_times()\n",
    "    ts_discretized = ts_tables.tree_sequence()\n",
    "    return(ts_discretized)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8994b371",
   "metadata": {},
   "outputs": [],
   "source": [
    "def impute_genotypes_using_tsinfer(ref_vcf_file,\n",
    "                                   miss_vcf_file,\n",
    "                                   imputed_vcf_file,\n",
    "                                   contig_id):\n",
    "    sd_ref  = create_sample_data_from_vcf_file(ref_vcf_file)\n",
    "    sd_miss = create_sample_data_from_vcf_file(miss_vcf_file)\n",
    "    ad_ref     = tsinfer.generate_ancestors(sample_data = sd_ref)\n",
    "    # This step is to infer a tree sequence from the sample data.\n",
    "    ts_anc_ref = tsinfer.match_ancestors(sample_data   = sd_ref,\n",
    "                                         ancestor_data = ad_ref)\n",
    "    ts_matched = tsinfer.match_samples(sample_data  = sd_miss,\n",
    "                                       ancestors_ts = ts_anc_ref)\n",
    "    with open(imputed_vcf_file, \"w\") as vcf:\n",
    "        ts_matched.write_vcf(vcf, contig_id = contig_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2bd1f1ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "def impute_genotypes_using_ts_only(ref_vcf_file,\n",
    "                                   miss_vcf_file,\n",
    "                                   imputed_vcf_file,\n",
    "                                   ts_anc_ref,\n",
    "                                   contig_id):\n",
    "    sd_ref  = create_sample_data_from_vcf_file(ref_vcf_file)\n",
    "    sd_miss = create_sample_data_from_vcf_file(miss_vcf_file)\n",
    "    ts_matched = tsinfer.match_samples(sample_data  = sd_miss,\n",
    "                                       ancestors_ts = ts_anc_ref)\n",
    "    with open(imputed_vcf_file, \"w\") as vcf:\n",
    "        ts_matched.write_vcf(vcf, contig_id = contig_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19a1ea4b",
   "metadata": {},
   "source": [
    "## Create data sets via simulations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "80ab9ee6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples from AMH : 100\n",
      "Number of samples from YRI : 250\n",
      "Number of samples from CEU : 9500\n",
      "Number of samples from CHB : 250\n",
      "Size of reference panel    : 10000\n"
     ]
    }
   ],
   "source": [
    "size_query =  1_000\n",
    "\n",
    "size_amh   =    100\n",
    "size_yri   =    250\n",
    "size_ceu   =  9_500\n",
    "size_chb   =    250\n",
    "size_ref   = size_yri + size_ceu + size_chb\n",
    "\n",
    "print(f\"Number of samples from AMH : {size_amh}\")\n",
    "print(f\"Number of samples from YRI : {size_yri}\")\n",
    "print(f\"Number of samples from CEU : {size_ceu}\")\n",
    "print(f\"Number of samples from CHB : {size_chb}\")\n",
    "print(f\"Size of reference panel    : {size_ref}\")\n",
    "\n",
    "mutation_rate = 1e-8\n",
    "recombination_rate = 1e-8\n",
    "\n",
    "num_replicates = 1\n",
    "\n",
    "proportion_missing_sites = 0.10\n",
    "num_missing_sites = 1_000\n",
    "\n",
    "contig_id = '1'\n",
    "ploidy_level = 1\n",
    "sequence_length = 1e7\n",
    "\n",
    "base_dir = \"../data/modern_ooa_unequal_900505_haploid_miss10/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f814f332",
   "metadata": {},
   "outputs": [],
   "source": [
    "rate_map = msprime.RateMap.uniform(\n",
    "    sequence_length = sequence_length,\n",
    "    rate = mutation_rate\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0c01a6b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (6)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>ancestral</td><td>Equilibrium/root population</td><td>0.0</td><td>0</td><td>8.8e+03</td><td>{}</td></tr><tr><td>1</td><td>AMH</td><td>Anatomically modern humans</td><td>0.0</td><td>0</td><td>5.6e+03</td><td>{}</td></tr><tr><td>2</td><td>OOA</td><td>Bottleneck out-of-Africa population</td><td>0.0</td><td>0</td><td>8.5e+02</td><td>{}</td></tr><tr><td>3</td><td>YRI</td><td>Yoruba in Ibadan, Nigeria</td><td>12300.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>4</td><td>CEU</td><td>Utah Residents (CEPH) with Northern and Western European Ancestry</td><td>29725.0</td><td>0.004</td><td>0</td><td>{}</td></tr><tr><td>5</td><td>CHB</td><td>Han Chinese in Beijing, China</td><td>54090.0</td><td>0.0055</td><td>0</td><td>{}</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Migration matrix</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th></th><th>ancestral</th><th>AMH</th><th>OOA</th><th>YRI</th><th>CEU</th><th>CHB</th></tr></thead><tbody><tr><td>ancestral</td><td>0</td><td title='Backwards in time migration rate from population ancestral to AMH = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from AMH to ancestral per generation, divided by the size of ancestral.'>0</td><td title='Backwards in time migration rate from population ancestral to OOA = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from OOA to ancestral per generation, divided by the size of ancestral.'>0</td><td title='Backwards in time migration rate from population ancestral to YRI = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from YRI to ancestral per generation, divided by the size of ancestral.'>0</td><td title='Backwards in time migration rate from population ancestral to CEU = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from CEU to ancestral per generation, divided by the size of ancestral.'>0</td><td title='Backwards in time migration rate from population ancestral to CHB = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from CHB to ancestral per generation, divided by the size of ancestral.'>0</td></tr><tr><td>AMH</td><td title='Backwards in time migration rate from population AMH to ancestral = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from ancestral to AMH per generation, divided by the size of AMH.'>0</td><td>0</td><td title='Backwards in time migration rate from population AMH to OOA = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from OOA to AMH per generation, divided by the size of AMH.'>0</td><td title='Backwards in time migration rate from population AMH to YRI = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from YRI to AMH per generation, divided by the size of AMH.'>0</td><td title='Backwards in time migration rate from population AMH to CEU = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from CEU to AMH per generation, divided by the size of AMH.'>0</td><td title='Backwards in time migration rate from population AMH to CHB = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from CHB to AMH per generation, divided by the size of AMH.'>0</td></tr><tr><td>OOA</td><td title='Backwards in time migration rate from population OOA to ancestral = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from ancestral to OOA per generation, divided by the size of OOA.'>0</td><td title='Backwards in time migration rate from population OOA to AMH = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from AMH to OOA per generation, divided by the size of OOA.'>0</td><td>0</td><td title='Backwards in time migration rate from population OOA to YRI = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from YRI to OOA per generation, divided by the size of OOA.'>0</td><td title='Backwards in time migration rate from population OOA to CEU = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from CEU to OOA per generation, divided by the size of OOA.'>0</td><td title='Backwards in time migration rate from population OOA to CHB = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from CHB to OOA per generation, divided by the size of OOA.'>0</td></tr><tr><td>YRI</td><td title='Backwards in time migration rate from population YRI to ancestral = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from ancestral to YRI per generation, divided by the size of YRI.'>0</td><td title='Backwards in time migration rate from population YRI to AMH = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from AMH to YRI per generation, divided by the size of YRI.'>0</td><td title='Backwards in time migration rate from population YRI to OOA = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from OOA to YRI per generation, divided by the size of YRI.'>0</td><td>0</td><td title='Backwards in time migration rate from population YRI to CEU = 3e-05 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from CEU to YRI per generation, divided by the size of YRI.'>3e-05</td><td title='Backwards in time migration rate from population YRI to CHB = 1.9e-05 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from CHB to YRI per generation, divided by the size of YRI.'>1.9e-05</td></tr><tr><td>CEU</td><td title='Backwards in time migration rate from population CEU to ancestral = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from ancestral to CEU per generation, divided by the size of CEU.'>0</td><td title='Backwards in time migration rate from population CEU to AMH = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from AMH to CEU per generation, divided by the size of CEU.'>0</td><td title='Backwards in time migration rate from population CEU to OOA = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from OOA to CEU per generation, divided by the size of CEU.'>0</td><td title='Backwards in time migration rate from population CEU to YRI = 3e-05 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from YRI to CEU per generation, divided by the size of CEU.'>3e-05</td><td>0</td><td title='Backwards in time migration rate from population CEU to CHB = 9.6e-05 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from CHB to CEU per generation, divided by the size of CEU.'>9.6e-05</td></tr><tr><td>CHB</td><td title='Backwards in time migration rate from population CHB to ancestral = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from ancestral to CHB per generation, divided by the size of CHB.'>0</td><td title='Backwards in time migration rate from population CHB to AMH = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from AMH to CHB per generation, divided by the size of CHB.'>0</td><td title='Backwards in time migration rate from population CHB to OOA = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from OOA to CHB per generation, divided by the size of CHB.'>0</td><td title='Backwards in time migration rate from population CHB to YRI = 1.9e-05 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from YRI to CHB per generation, divided by the size of CHB.'>1.9e-05</td><td title='Backwards in time migration rate from population CHB to CEU = 9.6e-05 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from CEU to CHB per generation, divided by the size of CHB.'>9.6e-05</td><td>0</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Events (16)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>848</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_parameters_change'>Population parameter change</a></td><td>population=OOA, initial_size=2100</td><td>initial_size → 2.1e+03 for population OOA</td></tr><tr><td>848</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[CEU, CHB], ancestral=OOA</td><td>Moves all lineages from derived populations &#x27;CEU&#x27; and &#x27;CHB&#x27; to the ancestral &#x27;OOA&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr><tr><td>848</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_migration_rate_change'>Migration rate change</a></td><td>source=CEU, dest=YRI, rate=0</td><td>Backwards-time migration rate from CEU to YRI → 0</td></tr><tr><td>848</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_migration_rate_change'>Migration rate change</a></td><td>source=YRI, dest=CEU, rate=0</td><td>Backwards-time migration rate from YRI to CEU → 0</td></tr><tr><td>848</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_migration_rate_change'>Migration rate change</a></td><td>source=CHB, dest=YRI, rate=0</td><td>Backwards-time migration rate from CHB to YRI → 0</td></tr><tr><td>848</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_migration_rate_change'>Migration rate change</a></td><td>source=YRI, dest=CHB, rate=0</td><td>Backwards-time migration rate from YRI to CHB → 0</td></tr><tr><td>848</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_migration_rate_change'>Migration rate change</a></td><td>source=CHB, dest=CEU, rate=0</td><td>Backwards-time migration rate from CHB to CEU → 0</td></tr><tr><td>848</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_migration_rate_change'>Migration rate change</a></td><td>source=CEU, dest=CHB, rate=0</td><td>Backwards-time migration rate from CEU to CHB → 0</td></tr><tr><td>848</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_migration_rate_change'>Migration rate change</a></td><td>source=OOA, dest=YRI, rate=0.00025</td><td>Backwards-time migration rate from OOA to YRI → 0.00025</td></tr><tr><td>848</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_migration_rate_change'>Migration rate change</a></td><td>source=YRI, dest=OOA, rate=0.00025</td><td>Backwards-time migration rate from YRI to OOA → 0.00025</td></tr><tr><td>5600</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_parameters_change'>Population parameter change</a></td><td>population=AMH, initial_size=12300</td><td>initial_size → 1.2e+04 for population AMH</td></tr><tr><td>5600</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[YRI, OOA], ancestral=AMH</td><td>Moves all lineages from derived populations &#x27;YRI&#x27; and &#x27;OOA&#x27; to the ancestral &#x27;AMH&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr><tr><td>5600</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_migration_rate_change'>Migration rate change</a></td><td>source=OOA, dest=YRI, rate=0</td><td>Backwards-time migration rate from OOA to YRI → 0</td></tr><tr><td>5600</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_migration_rate_change'>Migration rate change</a></td><td>source=YRI, dest=OOA, rate=0</td><td>Backwards-time migration rate from YRI to OOA → 0</td></tr><tr><td>8800</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_parameters_change'>Population parameter change</a></td><td>population=ancestral, initial_size=7300</td><td>initial_size → 7.3e+03 for population ancestral</td></tr><tr><td>8800</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[AMH], ancestral=ancestral</td><td>Moves all lineages from the &#x27;AMH&#x27; derived population to the ancestral &#x27;ancestral&#x27; population. Also set &#x27;AMH&#x27; to inactive, and all migration rates to and from the derived population to zero.</td></tr></tbody></table></div></div>"
      ],
      "text/plain": [
       "Demography(populations=[Population(initial_size=0, growth_rate=0, name='ancestral', description='Equilibrium/root population', extra_metadata={}, default_sampling_time=8800.0, initially_active=False, id=0), Population(initial_size=0, growth_rate=0, name='AMH', description='Anatomically modern humans', extra_metadata={}, default_sampling_time=5600.0, initially_active=False, id=1), Population(initial_size=0, growth_rate=0, name='OOA', description='Bottleneck out-of-Africa population', extra_metadata={}, default_sampling_time=848.0, initially_active=False, id=2), Population(initial_size=12300, growth_rate=0, name='YRI', description='Yoruba in Ibadan, Nigeria', extra_metadata={}, default_sampling_time=0.0, initially_active=True, id=3), Population(initial_size=29725, growth_rate=0.0039999863709632615, name='CEU', description='Utah Residents (CEPH) with Northern and Western European Ancestry', extra_metadata={}, default_sampling_time=0.0, initially_active=True, id=4), Population(initial_size=54090, growth_rate=0.005499992782013091, name='CHB', description='Han Chinese in Beijing, China', extra_metadata={}, default_sampling_time=0.0, initially_active=True, id=5)], events=[PopulationParametersChange(time=848.0, initial_size=2100, growth_rate=None, population='OOA'), PopulationSplit(time=848.0, derived=['CEU', 'CHB'], ancestral='OOA'), MigrationRateChange(time=848.0, rate=0, source='CEU', dest='YRI'), MigrationRateChange(time=848.0, rate=0, source='YRI', dest='CEU'), MigrationRateChange(time=848.0, rate=0, source='CHB', dest='YRI'), MigrationRateChange(time=848.0, rate=0, source='YRI', dest='CHB'), MigrationRateChange(time=848.0, rate=0, source='CHB', dest='CEU'), MigrationRateChange(time=848.0, rate=0, source='CEU', dest='CHB'), MigrationRateChange(time=848.0, rate=0.00025, source='OOA', dest='YRI'), MigrationRateChange(time=848.0, rate=0.00025, source='YRI', dest='OOA'), PopulationParametersChange(time=5600.0, initial_size=12300, growth_rate=None, population='AMH'), PopulationSplit(time=5600.0, derived=['YRI', 'OOA'], ancestral='AMH'), MigrationRateChange(time=5600.0, rate=0, source='OOA', dest='YRI'), MigrationRateChange(time=5600.0, rate=0, source='YRI', dest='OOA'), PopulationParametersChange(time=8800.0, initial_size=7300, growth_rate=None, population='ancestral'), PopulationSplit(time=8800.0, derived=['AMH'], ancestral='ancestral')], migration_matrix=array([[0.0e+00, 0.0e+00, 0.0e+00, 0.0e+00, 0.0e+00, 0.0e+00],\n",
       "       [0.0e+00, 0.0e+00, 0.0e+00, 0.0e+00, 0.0e+00, 0.0e+00],\n",
       "       [0.0e+00, 0.0e+00, 0.0e+00, 0.0e+00, 0.0e+00, 0.0e+00],\n",
       "       [0.0e+00, 0.0e+00, 0.0e+00, 0.0e+00, 3.0e-05, 1.9e-05],\n",
       "       [0.0e+00, 0.0e+00, 0.0e+00, 3.0e-05, 0.0e+00, 9.6e-05],\n",
       "       [0.0e+00, 0.0e+00, 0.0e+00, 1.9e-05, 9.6e-05, 0.0e+00]]))"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "yaml_file = \"../demes/gutenkunst_ooa_2009.yaml\"\n",
    "ooa_graph = demes.load(yaml_file)\n",
    "demography_model = msprime.Demography.from_demes(ooa_graph)\n",
    "demography_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4058b4a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Simulating 1 ts without duplicate site positions.\n",
      "Simulation of 1 ts took 0.99 seconds.\n"
     ]
    }
   ],
   "source": [
    "sample_set = [\n",
    "    msprime.SampleSet(num_samples = size_amh,\n",
    "                      population = \"AMH\", # id = 1\n",
    "                      ploidy = ploidy_level),\n",
    "    msprime.SampleSet(num_samples = size_query + size_yri,\n",
    "                      population = \"YRI\", # id = 3\n",
    "                      ploidy = ploidy_level),\n",
    "    msprime.SampleSet(num_samples = size_ceu,\n",
    "                      population = \"CEU\", # id = 4\n",
    "                      ploidy = ploidy_level),\n",
    "    msprime.SampleSet(num_samples = size_chb,\n",
    "                      population = \"CHB\", # id = 5\n",
    "                      ploidy = ploidy_level)\n",
    "]\n",
    "\n",
    "src_ts = [] # List of full simulated ts.\n",
    "\n",
    "tic = time.time()\n",
    "\n",
    "print(f\"Simulating {num_replicates} ts without duplicate site positions.\")\n",
    "success = 0\n",
    "\n",
    "while success < num_replicates:\n",
    "    sim_ts = msprime.sim_mutations(\n",
    "        msprime.sim_ancestry(\n",
    "            samples = sample_set,\n",
    "            demography = demography_model,\n",
    "            ploidy = ploidy_level,\n",
    "            model = \"hudson\",\n",
    "            recombination_rate = rate_map,\n",
    "            discrete_genome = True\n",
    "        ),\n",
    "        rate = mutation_rate,\n",
    "        discrete_genome = True\n",
    "    )\n",
    "    \n",
    "    src_ts.append(sim_ts)\n",
    "    success += 1\n",
    "    \n",
    "toc = time.time()\n",
    "print(f\"Simulation of {num_replicates} ts took {round(toc - tic, 2)} seconds.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a12ae6ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of ancient samples   : 100\n",
      "Number of query     samples : 1000\n",
      "Number of reference samples : 10000\n"
     ]
    }
   ],
   "source": [
    "# Impute into YRI samples.\n",
    "ts = src_ts[0]\n",
    "\n",
    "individuals_ancient = ts.samples(population = 1)\n",
    "samples_ancient     = individuals_ancient # When haploid\n",
    "\n",
    "individuals_query = ts.samples(population = 3)[:size_query]\n",
    "samples_query     = individuals_query # When haploid\n",
    "\n",
    "individuals_ref   = np.concatenate([ts.samples(population = 3)[size_query:],\n",
    "                                    ts.samples(population = 4),\n",
    "                                    ts.samples(population = 5)])\n",
    "samples_ref       = individuals_ref # When haploid\n",
    "\n",
    "gt_mask = mask_genotype.MissingGenotypeMask(individuals         = individuals_query,\n",
    "                                            sequence_length     = sequence_length,\n",
    "                                            proportion_missing  = proportion_missing_sites,\n",
    "                                            num_regions_missing = num_missing_sites,\n",
    "                                            contig_id           = contig_id)\n",
    "\n",
    "print(f\"Number of ancient samples   : {len(individuals_ancient)}\")\n",
    "print(f\"Number of query     samples : {len(individuals_query)}\")\n",
    "print(f\"Number of reference samples : {len(individuals_ref)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "fce9cca7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing ts 0.\n",
      "Printing ancient VCF.\n",
      "Printing query VCF with non-missing genotypes.\n"
     ]
    }
   ],
   "source": [
    "anc_ts = [] # List of simulated ancestor ts.\n",
    "\n",
    "for i, ts in enumerate(src_ts):\n",
    "    print(f\"Processing ts {i}.\")\n",
    "    ancient_vcf_file  = base_dir + \"ancient/\"  + \"ancient.\"  + str(i) + \".vcf\"\n",
    "    ref_vcf_file  = base_dir + \"ref/\"  + \"ref.\"  + str(i) + \".vcf\"\n",
    "    true_vcf_file = base_dir + \"true/\" + \"true.\" + str(i) + \".vcf\"\n",
    "    miss_vcf_file = base_dir + \"miss/\" + \"miss.\" + str(i) + \".vcf\"\n",
    "    ts_anc_ref_file = base_dir + \"ts_anc_ref/\" + \"ts_anc_ref.\" + str(i) + \".trees\"\n",
    "    \n",
    "    sd_all = tsinfer.SampleData.from_tree_sequence(ts, use_sites_time = False)\n",
    "    sd_ancient = sd_all.subset(individuals = individuals_ancient)\n",
    "    sd_ref     = sd_all.subset(individuals = individuals_ref)\n",
    "    sd_query   = sd_all.subset(individuals = individuals_query)\n",
    "    \n",
    "    sites_to_keep     = find_biallelic_sites(sd_ref, sd_query)\n",
    "    sd_ref_filtered   =   sd_ref.subset(sites = sites_to_keep[0])\n",
    "    sd_query_filtered = sd_query.subset(sites = sites_to_keep[1])\n",
    "    \n",
    "    # TODO: Refactor.\n",
    "    #print(\"Printing ancestors ts.\")\n",
    "    #sim_ts_anc_ref = make_ancestors_ts(samples = samples_ref,\n",
    "    #                                   ts = ts,\n",
    "    #                                   remove_leaves = True)\n",
    "    #tmp_tables = sim_ts_anc_ref.dump_tables()\n",
    "    #tmp_tables.populations.metadata_schema = tskit.MetadataSchema(schema = None)\n",
    "    #sim_ts_anc_ref = tmp_tables.tree_sequence()\n",
    "    #anc_ts.append(sim_ts_anc_ref)\n",
    "    #sim_ts_anc_ref.dump(ts_anc_ref_file)\n",
    "    \n",
    "    print(\"Printing ancient VCF.\")\n",
    "    print_sample_data_to_vcf(sample_data = sd_ancient,\n",
    "                             individuals = individuals_ancient,\n",
    "                             samples = samples_ancient,\n",
    "                             ploidy_level = ploidy_level,\n",
    "                             mask = None,\n",
    "                             out_vcf_file = ancient_vcf_file,\n",
    "                             contig_id = contig_id,\n",
    "                             sequence_length_max = 1e24)\n",
    "    \n",
    "    #print(\"Printing reference panel VCF.\")\n",
    "    #print_sample_data_to_vcf(sample_data = sd_ref_filtered,\n",
    "    #                         individuals = individuals_ref,\n",
    "    #                         samples = samples_ref,\n",
    "    #                         ploidy_level = ploidy_level,\n",
    "    #                         mask = None,\n",
    "    #                         out_vcf_file = ref_vcf_file,\n",
    "    #                         contig_id = contig_id,\n",
    "    #                         sequence_length_max = 1e24)\n",
    "    \n",
    "    print(\"Printing query VCF with non-missing genotypes.\")\n",
    "    print_sample_data_to_vcf(sample_data = sd_query_filtered,\n",
    "                             individuals = individuals_query,\n",
    "                             samples = samples_query,\n",
    "                             ploidy_level = ploidy_level,\n",
    "                             mask = None,\n",
    "                             out_vcf_file = true_vcf_file,\n",
    "                             contig_id = contig_id,\n",
    "                             sequence_length_max = 1e24)\n",
    "    \n",
    "    #print(\"Printing query VCF with missing genotypes.\")\n",
    "    #print_sample_data_to_vcf(sample_data = sd_query_filtered,\n",
    "    #                         individuals = individuals_query,\n",
    "    #                         samples = samples_query,\n",
    "    #                         ploidy_level = ploidy_level,\n",
    "    #                         mask = gt_mask,\n",
    "    #                         out_vcf_file = miss_vcf_file,\n",
    "    #                         contig_id = contig_id,\n",
    "    #                         sequence_length_max = 1e24)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6da1bc5",
   "metadata": {},
   "source": [
    "## Perform genotype imputation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "3f2d72ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing imputation using ts only.\n",
      "Imputing VCF 0\n"
     ]
    }
   ],
   "source": [
    "print(\"Doing imputation using ts only.\")\n",
    "\n",
    "for i in np.arange(len(src_ts)):\n",
    "    print(f\"Imputing VCF {i}\")\n",
    "    ref_vcf_file     = base_dir + \"ref/\"  + \"ref.\"  + str(i) + \".vcf\"\n",
    "    miss_vcf_file    = base_dir + \"miss/\" + \"miss.\" + str(i) + \".vcf\"\n",
    "    imputed_vcf_file = base_dir + \"imputed_tsonly/\" + \"imputed.\" + str(i) + \".vcf\"\n",
    "    impute_genotypes_using_ts_only(ref_vcf_file = ref_vcf_file,\n",
    "                                   miss_vcf_file = miss_vcf_file,\n",
    "                                   imputed_vcf_file = imputed_vcf_file,\n",
    "                                   ts_anc_ref = anc_ts[i],\n",
    "                                   contig_id = contig_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e143b872",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing imputation using tsinfer.\n",
      "Imputing VCF 0\n"
     ]
    }
   ],
   "source": [
    "print(\"Doing imputation using tsinfer.\")\n",
    "\n",
    "#for i in np.arange(len(src_ts)):\n",
    "for i in np.arange(1):\n",
    "    print(f\"Imputing VCF {i}\")\n",
    "    ref_vcf_file     = base_dir + \"ref/\" + \"ref.\"  + str(i) + \".vcf\"\n",
    "    miss_vcf_file    = base_dir + \"miss/\" + \"miss.\" + str(i) + \".vcf\"\n",
    "    imputed_vcf_file = base_dir + \"imputed_tsinfer/\" + \"imputed.\" + str(i) + \".vcf\"\n",
    "    impute_genotypes_using_tsinfer(ref_vcf_file = ref_vcf_file,\n",
    "                                   miss_vcf_file = miss_vcf_file,\n",
    "                                   imputed_vcf_file = imputed_vcf_file,\n",
    "                                   contig_id = contig_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "ff859033",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing imputation using BEAGLE.\n",
      "Imputing VCF 0\n",
      "java -jar ../analysis/beagle/beagle.28Jun21.220.jar ref=../data/modern_ooa_unequal_900505_haploid_miss10/ref/ref.0.vcf gt=../data/modern_ooa_unequal_900505_haploid_miss10/miss/miss.0.vcf out=../data/modern_ooa_unequal_900505_haploid_miss10/imputed_beagle/imputed.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"Doing imputation using BEAGLE.\")\n",
    "\n",
    "beagle_exe = \"../analysis/beagle/beagle.28Jun21.220.jar\"\n",
    "\n",
    "for i in np.arange(len(src_ts)):\n",
    "    print(f\"Imputing VCF {i}\")\n",
    "    ref_vcf_file     = base_dir + \"ref/\"  + \"ref.\"  + str(i) + \".vcf\"\n",
    "    miss_vcf_file    = base_dir + \"miss/\" + \"miss.\" + str(i) + \".vcf\"\n",
    "    imputed_vcf_file = base_dir + \"imputed_beagle/\" + \"imputed.\" + str(i)\n",
    "    beagle_cmd = [\n",
    "        \"java\", \"-jar\", beagle_exe,\n",
    "        \"ref=\" + ref_vcf_file,\n",
    "        \"gt=\"  + miss_vcf_file,\n",
    "        \"out=\" + imputed_vcf_file\n",
    "    ]\n",
    "    beagle_cmd = \" \".join(beagle_cmd)\n",
    "    print(beagle_cmd + \"\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
